pagefaultgames/pokerogue
# [Dev] Bump Game Version, Overhaul Version Migration

### Frontend companion PR to pagefaultgames/rogueserver#55





## What are the changes the user will see?


Sessions around version `1.0.0` had battle items with higher `battleCounts` than `maxBattles`, causing a minor issue with `battleCount` coloring and having more battles left than intended, which has been in fixed. Besides this, no other changes should be seen besides the version changing from `1.0.4` to `1.1.0`.

## Why am I making these changes?




Version migration was put together in a way that was quick but not easily maintainable for future changes or patch additions, since line bloat would easily take over the file as time goes on. Additionally, there was a halt on versioning at all given that the server would force the client to reload on receiving outdated sessions, which is no longer an issue with real-time version migration. With the companion server changes, versioning can also now be resumed and migration can be done without repeats (causing scenarios like #4100).

## What are the changes from a developer perspective?



- Utilizes [`npm version minor --force --no-git-tag-version`](https://docs.npmjs.com/cli/v9/commands/npm-version) to increment the version up to `1.1.0`.
- Moved all patches into their own separate files all under the `versions` folder
  - e.g, all `1.0.4` and below patches exist in `src/system/version_migration/versions/v1_0_4.ts` with specific arrays per each data type (`sessionMigrators`, `systemMigrators`, etc.)
- `src/system/version_migration/version_converter.ts` now holds several `VersionConverter` classes that better reuse and encapsulate version parsing, comparison, and migration.
  - Each `VersionConverter` supports application of functions at a granular level with differentiation based on major, minor, and patch values in addition to having different patching timelines. This means that `SessionSaveData` could have a lot of different version specific patches migrating from `1.0.4` to `1.1.0` to `1.X.X` with multiple `case` entries within the existing `switch` blocks, but that doesn't need to be the case for `SystemSaveData` since it has its own `applyMigration` function. 

### Screenshots/Videos



For reference, this is the phase that was removed.

<img width="627" alt="outdatedPhase" src="https://github.com/user-attachments/assets/7e5dae0c-6b68-4f57-bd37-dc43345e5edf">

## How to test the changes?





If on hand, any save data or session data can be imported to check if the game will successfully migrate from its previous version to the latest version without hiccup. 

For the server changes, a local mirror of the server using Docker connected to the frontend would need to be installed in order to see that newer versions beyond `1.0.4` will load without causing error or flagging the removed `OutdatedPhase`.

## Checklist
  -  https://github.com/pagefaultgames/pokerogue/pull/4344
